can/interfaces/nixnet.py

```diff
@@ -0,0 +1,257 @@
+        """
+        :param str channel:
+            Name of the object to open (e.g. 'CAN0')
+
+        :param int bitrate:
+            Bitrate in bits/s
+
+        :param list can_filters:
+            See :meth:`can.BusABC.set_filters`.
+
+        :param bool log_errors:
+            If True, communication errors will appear as CAN messages with
+            ``is_error_frame`` set to True and ``arbitration_id`` will identify
+            the error (default True)
+
+        :raises can.interfaces.nixnet.NiXNETError:
+            If starting communication fails
+
+        """
+        self._rx_queue = []
+        self.channel = channel
+        self.channel_info = "NI-XNET: " + channel
+
+        # Set database for the initialization
+        if not fd:
+            database_name = ":memory:"
+        else:
+            if not brs:
+                database_name = ":can_fd:"
+            else:
+                database_name = ":can_fd_brs:"
+
+        try:
+
+            # We need two sessions for this application, one to send frames and another to receive them
+
+            self.__session_send = session.FrameOutStreamSession(
+                channel, database_name=database_name
+            )
+            self.__session_receive = session.FrameInStreamSession(
+                channel, database_name=database_name
+            )
+
+            # We stop the sessions to allow reconfiguration, as by default they autostart at creation
+            self.__session_send.stop()
+            self.__session_receive.stop()
+
+            # See page 1017 of NI-XNET Hardware and Software Manual to set custom can configuration
+            if bitrate:
+                self.__session_send.intf.baud_rate = bitrate
+                self.__session_receive.intf.baud_rate = bitrate
+
+            if fd_bitrate:
+                # See page 951 of NI-XNET Hardware and Software Manual to set custom can configuration
+                self.__session_send.intf.can_fd_baud_rate = fd_bitrate
+                self.__session_receive.intf.can_fd_baud_rate = fd_bitrate
+
+            if can_termination:
+                self.__session_send.intf.can_term = constants.CanTerm.ON
+                self.__session_receive.intf.can_term = constants.CanTerm.ON
+
+            self.__session_receive.queue_size = 512
+            # Once that all the parameters have been restarted, we start the sessions
+            self.__session_send.start()
+            self.__session_receive.start()
+
+        except errors.XnetError as err:
+            raise NiXNETError(function="__init__", error_message=err.args[0]) from None
+
+        self._is_filtered = False
+        super(NiXNETcanBus, self).__init__(
+            channel=channel,
+            can_filters=can_filters,
+            bitrate=bitrate,
+            log_errors=log_errors,
+            **kwargs
+        )
+
+    def _recv_internal(self, timeout):
+        try:
+            fr = self.__session_receive.frames.read(4, timeout=0.0)
```

Comment:
This is only a trick to make it working without modifying the nixnet library from National. With a positive timeout of any value it doesn't works. And regarding the queue and the 4 frames reading is because they have hardcoded a fixed length for reading of 24 bytes, so this is needed to be able to read 64 bytes CANFD frames, or at least, is the easy way I have found without patching the underlaying library. Also, it's a good idea to skip the reading if the queue is not empty, I'll do it.

Reply:
The library definitely seems a bit unfinished but that's not your fault.